lib/annotations2triannon/open_annotation.rb

```diff
@@ -261,6 +254,33 @@ def to_ttl
       @graph.dump(:ttl, standard_prefixes: true)
     end
 
+    private
+
+    # Sets the annotation ID as an RDF::URI from id, but if the id is nil, it
+    # will try to get the ID from the graph and, as a last resort, it will
+    # generate a unique ID using a UUID.
+    # @param id [UUID|URI|String] to identify an open annotation
+    def set_id(id)
+      raise ArgumentError, 'id cannot be an empty String' if (id.instance_of?(String) && id.empty?)
+      if id.nil?
+        # Try to get the ID from the graph or use a UUID
+        q = [nil, RDF.type, OA.Annotation]
+        @id = @graph.query(q).collect {|s| s.subject }.first || RDF::URI.parse(UUID.generate)
```

Comment:
I don't think we need an integration test, just to verify manually that the
interactions are as expected.
On May 19, 2015 10:45 AM, "Darren L. Weber, Ph.D." <notifications@github.com>
wrote:

> In lib/annotations2triannon/open_annotation.rb
> <https://github.com/sul-dlss/annotations2triannon/pull/20#discussion_r30625801>
> :
>
> > @@ -261,6 +254,33 @@ def to_ttl
> >        @graph.dump(:ttl, standard_prefixes: true)
> >      end
> >
> > +    private
> > +
> > +    # Sets the annotation ID as an RDF::URI from id, but if the id is nil, it
> > +    # will try to get the ID from the graph and, as a last resort, it will
> > +    # generate a unique ID using a UUID.
> > +    # @param id [UUID|URI|String] to identify an open annotation
> > +    def set_id(id)
> > +      raise ArgumentError, 'id cannot be an empty String' if (id.instance_of?(String) && id.empty?)
> > +      if id.nil?
> > +        # Try to get the ID from the graph or use a UUID
> > +        q = [nil, RDF.type, OA.Annotation]
> > +        @id = @graph.query(q).collect {|s| s.subject }.first || RDF::URI.parse(UUID.generate)
>
> OK, I think I'm following this, but there are lots of unanswered questions
> and @azaroth42 <https://github.com/azaroth42> suggests we create an
> integration test to explore the client<->triannon and triannon<->FC4
> interactions (hopefully Rob will manage work for that).
>
> Anyhow, I think the main point for this PR is that
> RDF::URI.new(UUID.generate) or RDF::Node.new are reasonable alternatives.
> The code is designed to use a rule to define the default, rather than a
> simple default. This will be explained further in additional code comments
> (in an addendum commit to this PR, coming soon). At a later stage (i.e.
> TODO item), there could be a config parameter to use one or the other as a
> default.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sul-dlss/annotations2triannon/pull/20/files#r30625801>
> .
>